core/src/execute_cost_table.rs

```diff
@@ -0,0 +1,274 @@
+            occurrences: HashMap::new(),
+        }
+    }
+
+    pub fn get_cost_table(&self) -> &HashMap<Pubkey, u64> {
+        &self.table
+    }
+
+    pub fn get_count(&self) -> usize {
+        self.table.len()
+    }
+
+    // instead of assigning unknown program with a configured/hard-coded cost
+    // use average or mode function to make a educated guess.
+    pub fn get_average(&self) -> u64 {
+        if self.table.is_empty() {
+            0
+        } else {
+            self.table.iter().map(|(_, value)| value).sum::<u64>() / self.get_count() as u64
+        }
+    }
+
+    pub fn get_mode(&self) -> u64 {
+        if self.occurrences.is_empty() {
+            0
+        } else {
+            let key = self
+                .occurrences
+                .iter()
+                .max_by_key(|&(_, count)| count)
+                .map(|(key, _)| key)
+                .expect("cannot find mode from cost table");
+
+            *self.table.get(&key).unwrap()
+        }
+    }
+
+    // returns None if program doesn't exist in table. In this case,
+    // client is advised to call `get_average()` or `get_mode()` to
+    // assign a 'default' value for new program.
+    pub fn get_cost(&self, key: &Pubkey) -> Option<&u64> {
+        self.table.get(&key)
+    }
+
+    pub fn upsert(&mut self, key: &Pubkey, value: &u64) {
+        let need_to_add = self.table.get(&key).is_none();
+        let current_size = self.get_count();
+        if current_size == self.capacity && need_to_add {
+            self.prune_to(&((current_size as f64 * PRUNE_RATIO) as usize));
+        }
+
+        let program_cost = self.table.entry(*key).or_insert(*value);
+        *program_cost = (*program_cost + *value) / 2;
+
+        let (count, timestamp) = self
+            .occurrences
+            .entry(*key)
+            .or_insert((0, SystemTime::now()));
+        *count += 1;
+        *timestamp = SystemTime::now();
+    }
+
+    // prune the old programs so the table contains `new_size` of records,
+    // where `old` is defined as weighted age, which is negatively correlated
+    // with program's age and
+    // positively correlated with how frequently the program
+    // is executed (eg. occurrence),
+    fn prune_to(&mut self, new_size: &usize) {
+        debug!(
+            "prune cost table, current size {}, new size {}",
+            self.get_count(),
+            new_size
+        );
+
+        if *new_size == self.get_count() {
+            return;
+        }
+
+        if *new_size == 0 {
+            self.table.clear();
+            self.occurrences.clear();
+            return;
+        }
+
+        let now = SystemTime::now();
+        let mut sorted_by_weighted_age: Vec<_> = self
+            .occurrences
+            .iter()
+            .map(|(key, (count, timestamp))| {
+                let weighted_age: f64 =
+                    *count as f64 / (now.duration_since(*timestamp).unwrap().as_micros()) as f64;
```

Comment:
potential to divide by 0 here?

Reply:
yea, dont like division anyway, just pushed a commit changing formula to `=count * WEIGHT + (-1)*age`